File: api/routes/trading.py
"""REST API routes for paper trading tasks."""
from __future__ import annotations

import uuid
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request

from api.db_ext import (
    save_paper_trading_task,
    get_paper_trading_task,
    update_paper_trading_task,
    list_paper_trading_tasks,
    count_paper_trading_tasks,
    list_paper_trades,
    count_paper_trades,
    delete_paper_trading_task,
    list_equity_snapshots,
    compute_trading_metrics,
)
from api.deps import get_db_path
from api.schemas import (
    PaperTradingTaskCreate,
    PaperTradingTaskResponse,
    PaperTradingTaskListResponse,
    PaperTradeListResponse,
    PaperTradeResponse,
    EquitySnapshotResponse,
    EquitySnapshotListResponse,
    TradingMetricsResponse,
)

router = APIRouter(prefix="/api/trading", tags=["trading"])


def _task_to_response(row: dict) -> PaperTradingTaskResponse:
    return PaperTradingTaskResponse(
        task_id=row["task_id"],
        status=row["status"],
        strategy_name=row.get("strategy_name"),
        symbol=row["symbol"],
        timeframe=row["timeframe"],
        initial_cash=row["initial_cash"],
        fee=row["fee"],
        leverage=row["leverage"],
        direction=row["direction"],
        score_template=row.get("score_template", "explorer"),
        created_at=row["created_at"],
        updated_at=row["updated_at"],
        started_at=row.get("started_at"),
        stopped_at=row.get("stopped_at"),
        stop_reason=row.get("stop_reason"),
        position_side=row.get("position_side"),
        position_entry=row.get("position_entry"),
        position_quantity=row.get("position_quantity"),
        position_margin=row.get("position_margin"),
        position_funding=row.get("position_funding"),
        balance=row.get("balance"),
        unrealized_pnl=row.get("unrealized_pnl", 0.0),
        total_trades=row.get("total_trades", 0),
        total_pnl=row.get("total_pnl", 0.0),
        win_count=row.get("win_count", 0),
        loss_count=row.get("loss_count", 0),
        last_bar_time=row.get("last_bar_time"),
        last_bar_close=row.get("last_bar_close"),
        execution_model=row.get("execution_model", "v1"),
        confidence_sizing_enabled=bool(row.get("confidence_sizing_enabled", 0)),
        prediction_dna_json=row.get("prediction_dna_json"),
    )


@router.post("/tasks", response_model=PaperTradingTaskResponse, status_code=201)
def create_task(
    body: PaperTradingTaskCreate,
    db_path: Path = Depends(get_db_path),
) -> PaperTradingTaskResponse:
    # Validate dna_json format
    try:
        from core.strategy.dna import StrategyDNA
        StrategyDNA.from_json(body.dna_json)
    except Exception as e:
        raise HTTPException(status_code=422, detail=f"Invalid dna_json: {e}")

    task_id = uuid.uuid4().hex[:12]
    save_paper_trading_task(
        db_path,
        task_id=task_id,
        dna_json=body.dna_json,
        symbol=body.symbol,
        timeframe=body.timeframe,
        initial_cash=body.initial_cash,
        fee=body.fee,
        leverage=body.leverage,
        direction=body.direction,
        score_template=body.score_template,
        strategy_name=body.strategy_name,
        confidence_sizing_enabled=body.confidence_sizing_enabled,
        prediction_dna_json=body.prediction_dna_json,
    )
    row = get_paper_trading_task(db_path, task_id)
    response = _task_to_response(row)

    # Qualified warning check: if strategy_id provided, look up qualified status
    warnings = []
    if body.strategy_id:
        from api.db_ext import get_strategy
        strat = get_strategy(db_path, body.strategy_id)
        if strat is not None:
            qualified_val = strat.get("qualified")
            if qualified_val is not None and not bool(qualified_val):
                warnings.append(
                    "Strategy has not passed qualification checks (qualified=False). "
                    "Performance may not meet minimum requirements."
                )

    response.warnings = warnings if warnings else None
    return response


@router.get("/tasks", response_model=PaperTradingTaskListResponse)
def list_tasks(
    status: Optional[str] = None,
    limit: int = 50,
    offset: int = 0,
    db_path: Path = Depends(get_db_path),
) -> PaperTradingTaskListResponse:
    tasks = list_paper_trading_tasks(db_path, status=status, limit=limit, offset=offset)
    total = count_paper_trading_tasks(db_path, status=status)
    return PaperTradingTaskListResponse(
        tasks=[_task_to_response(t) for t in tasks],
        total=total,
    )


@router.get("/tasks/{task_id}", response_model=PaperTradingTaskResponse)
def get_task(
    task_id: str,
    db_path: Path = Depends(get_db_path),
) -> PaperTradingTaskResponse:
    row = get_paper_trading_task(db_path, task_id)
    if row is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return _task_to_response(row)


@router.post("/tasks/{task_id}/stop", response_model=PaperTradingTaskResponse)
def stop_task(
    task_id: str,
    db_path: Path = Depends(get_db_path),
) -> PaperTradingTaskResponse:
    row = get_paper_trading_task(db_path, task_id)
    if row is None:
        raise HTTPException(status_code=404, detail="Task not found")
    if row["status"] not in ("running", "pending"):
        raise HTTPException(status_code=400, detail=f"Cannot stop task in status '{row['status']}'")

    # Signal via controller (immediate)
    from core.trading.runner import get_trading_controllers
    controller = get_trading_controllers().get(task_id)
    if controller:
        controller.request_stop()

    # Also set DB status (belt-and-suspenders)
    update_paper_trading_task(db_path, task_id, status="stopped", stop_reason="user_stop")
    row = get_paper_trading_task(db_path, task_id)
    return _task_to_response(row)


@router.post("/tasks/{task_id}/restart", response_model=PaperTradingTaskResponse)
def restart_task(
    task_id: str,
    db_path: Path = Depends(get_db_path),
) -> PaperTradingTaskResponse:
    """Create a new task using the same strategy/params as an existing task."""
    row = get_paper_trading_task(db_path, task_id)
    if row is None:
        raise HTTPException(status_code=404, detail="Task not found")

    new_task_id = uuid.uuid4().hex[:12]
    save_paper_trading_task(
        db_path,
        task_id=new_task_id,
        strategy_name=row.get("strategy_name"),
        symbol=row["symbol"],
        timeframe=row["timeframe"],
        initial_cash=row["initial_cash"],
        fee=row["fee"],
        leverage=row["leverage"],
        direction=row["direction"],
        dna_json=row["dna_json"],
        score_template=row.get("score_template", "explorer"),
        confidence_sizing_enabled=bool(row.get("confidence_sizing_enabled", 0)),
        prediction_dna_json=row.get("prediction_dna_json"),
    )
    new_row = get_paper_trading_task(db_path, new_task_id)
    return _task_to_response(new_row)


@router.get("/tasks/{task_id}/trades", response_model=PaperTradeListResponse)
def get_trades(
    task_id: str,
    limit: int = 100,
    db_path: Path = Depends(get_db_path),
) -> PaperTradeListResponse:
    row = get_paper_trading_task(db_path, task_id)
    if row is None:
        raise HTTPException(status_code=404, detail="Task not found")
    trades = list_paper_trades(db_path, task_id, limit=limit)
    total = count_paper_trades(db_path, task_id)
    return PaperTradeListResponse(
        trades=[PaperTradeResponse(**t) for t in trades],
        total=total,
    )


@router.get("/tasks/{task_id}/equity", response_model=EquitySnapshotListResponse)
def get_equity(
    task_id: str,
    db_path: Path = Depends(get_db_path),
) -> EquitySnapshotListResponse:
    row = get_paper_trading_task(db_path, task_id)
    if row is None:
        raise HTTPException(status_code=404, detail="Task not found")
    snapshots = list_equity_snapshots(db_path, task_id)
    return EquitySnapshotListResponse(
        snapshots=[EquitySnapshotResponse(**s) for s in snapshots],
        total=len(snapshots),
    )


@router.get("/tasks/{task_id}/metrics", response_model=TradingMetricsResponse)
def get_metrics(
    task_id: str,
    db_path: Path = Depends(get_db_path),
) -> TradingMetricsResponse:
    metrics = compute_trading_metrics(db_path, task_id)
    if metrics is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return TradingMetricsResponse(**metrics)


@router.delete("/tasks/{task_id}")
def delete_task(
    task_id: str,
    db_path: Path = Depends(get_db_path),
) -> dict:
    row = get_paper_trading_task(db_path, task_id)
    if row is None:
        raise HTTPException(status_code=404, detail="Task not found")
    if row["status"] in ("running", "pending"):
        raise HTTPException(status_code=400, detail="Cannot delete active task. Stop it first.")
    ok = delete_paper_trading_task(db_path, task_id)
    return {"deleted": ok}


@router.get("/runner-status")
def runner_status(request: Request) -> dict:
    runner = getattr(request.app.state, "trading_runner", None)
    if runner:
        return runner.get_status()
    return {"is_alive": False, "active_task_id": None}

File: web/src/services/trading.ts
import { api } from "./api";

export interface TradingTask {
  task_id: string;
  status: string;
  strategy_name: string | null;
  symbol: string;
  timeframe: string;
  initial_cash: number;
  fee: number;
  leverage: number;
  direction: string;
  score_template: string;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  stopped_at: string | null;
  stop_reason: string | null;
  position_side: string | null;
  position_entry: number | null;
  position_quantity: number | null;
  position_margin: number | null;
  position_funding: number | null;
  balance: number | null;
  unrealized_pnl: number;
  total_trades: number;
  total_pnl: number;
  win_count: number;
  loss_count: number;
  last_bar_time: string | null;
  last_bar_close: number | null;
  bars_held: number;
}

export interface TradingTaskList {
  tasks: TradingTask[];
  total: number;
}

export interface PaperTrade {
  id: number;
  task_id: string;
  bar_time: string;
  side: string;
  action: string;
  price: number;
  quantity: number;
  pnl: number | null;
  fee_paid: number;
  reason: string | null;
}

export interface PaperTradeList {
  trades: PaperTrade[];
  total: number;
}

export interface CreateTradingTaskParams {
  dna_json: string;
  symbol?: string;
  timeframe?: string;
  initial_cash?: number;
  fee?: number;
  leverage?: number;
  direction?: string;
  score_template?: string;
  strategy_name?: string;
}

export async function listTradingTasks(params?: {
  status?: string;
  limit?: number;
  offset?: number;
}): Promise<TradingTaskList> {
  const { data } = await api.get("/api/trading/tasks", { params });
  return data;
}

export async function getTradingTask(taskId: string): Promise<TradingTask> {
  const { data } = await api.get(`/api/trading/tasks/${taskId}`);
  return data;
}

export async function createTradingTask(
  params: CreateTradingTaskParams
): Promise<TradingTask> {
  const { data } = await api.post("/api/trading/tasks", params);
  return data;
}

export async function stopTradingTask(taskId: string): Promise<TradingTask> {
  const { data } = await api.post(`/api/trading/tasks/${taskId}/stop`);
  return data;
}

export async function pauseTradingTask(taskId: string): Promise<TradingTask> {
  const { data } = await api.post(`/api/trading/tasks/${taskId}/pause`);
  return data;
}

export async function resumeTradingTask(taskId: string): Promise<TradingTask> {
  const { data } = await api.post(`/api/trading/tasks/${taskId}/resume`);
  return data;
}

export async function restartTradingTask(taskId: string): Promise<TradingTask> {
  const { data } = await api.post(`/api/trading/tasks/${taskId}/restart`);
  return data;
}

export async function getTradingTrades(
  taskId: string,
  limit?: number
): Promise<PaperTradeList> {
  const { data } = await api.get(`/api/trading/tasks/${taskId}/trades`, {
    params: { limit },
  });
  return data;
}

export async function getTradingRunnerStatus(): Promise<{
  is_alive: boolean;
  active_task_id: string | null;
}> {
  const { data } = await api.get("/api/trading/runner-status");
  return data;
}

export interface EquitySnapshot {
  id: number;
  task_id: string;
  timestamp: string;
  equity: number;
  balance: number;
  unrealized_pnl: number;
  position_side: string;
}

export interface EquitySnapshotList {
  snapshots: EquitySnapshot[];
  total: number;
}

export interface TradingMetrics {
  task_id: string;
  total_return: number;
  total_return_pct: number;
  win_rate: number;
  profit_factor: number | null;
  max_drawdown: number;
  max_drawdown_pct: number;
  avg_trade_pnl: number;
  total_trades: number;
  total_pnl: number;
  win_count: number;
  loss_count: number;
}

export async function getTradingEquity(taskId: string): Promise<EquitySnapshotList> {
  const { data } = await api.get(`/api/trading/tasks/${taskId}/equity`);
  return data;
}

export async function getTradingMetrics(taskId: string): Promise<TradingMetrics> {
  const { data } = await api.get(`/api/trading/tasks/${taskId}/metrics`);
  return data;
}

export async function deleteTradingTask(taskId: string): Promise<{ deleted: boolean }> {
  const { data } = await api.delete(`/api/trading/tasks/${taskId}`);
  return data;
}
